### DISCUSSION TITLE
Wrong order of path paramenters in openapi.json, when dependencies are being used

### DISCUSSION THREAD

**Discussion by poul1x** [fastapi/fastapi#8600] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from dataclasses import dataclass
from http.client import HTTPException
from fastapi import APIRouter, Depends, FastAPI, Request

async def check_current_user(request: Request, user_id: str):
    # Some logic, check user exists
    request.state.current_user_id = user_id


@dataclass
class Project:
    name: str
    descrition: str


projects = {
    "project1": Project("Dev", "Development"),
    "project2": Project("Prod", "Production"),
    "project3": Project("Test", "Testing"),
}


async def check_parent_project(request: Request, project_id: str):
    if project_id not in projects.keys():
        raise HTTPException("project_id not in projects_ids")
    request.state.project = projects[project_id]


async def parent_project(request: Request):
    return request.state.project


app = FastAPI()
router_users = APIRouter(prefix="/users/{user_id}")

router_projects = APIRouter(
    prefix="/projects/{project_id}",
    dependencies=[
        Depends(check_current_user),
    ],
)

router_issues = APIRouter(
    prefix="/issues/{issue_id}",
    dependencies=[Depends(check_parent_project)],
)


@router_users.get("")
async def get_user(user_id: str):
    return {"message": f"Hello user {user_id}"}


@router_projects.get("")
async def get_project(project_id: str):
    return {"message": f"Hello user, your project is {project_id}"}


@router_issues.get("/no-depends")
async def get_issue(user_id: str, project_id: str, issue_id: str):
    return {
        "message": f"Hello user {user_id}, your project is {project_id}, issue is {issue_id}"
    }


@router_issues.get("")
async def get_issue(
    issue_id: str,
    project: Project = Depends(parent_project),
):
    return {
        "message": f"Hello user, your project is {project.name}, your issue is {issue_id}"
    }


router_projects.include_router(router_issues)
router_users.include_router(router_projects)
app.include_router(router_users)
```

### Description

I'm using dependencies to transparently check that top level resource exists, before the main logic of route handler begins.
Unfortunately, I've noticed, that my code messes up `openapi.json`: `project_id` is placed before `user_id`, but logically shouldn't. This situation is really annoing, when automatic code generation is being used. I have many functions with out-of-order arguments.

<img src="https://user-images.githubusercontent.com/31695750/155021753-7fa35ae7-7fcf-400e-b839-101e5d94162a.png" height="300px">

There's another route without dependencies. It's logically ok.

<img src="https://user-images.githubusercontent.com/31695750/155021922-d6cfce3c-59af-450e-a8c0-73eef5e16288.png" height="350px">

Questions:
1. Is it normal behaviour for FastAPI?
2. Any hacks to fix `openapi.json`?

### Operating System

Windows

### FastAPI Version

0.73.0

### Python Version

3.9.2

**Comment by poul1x** [comment]
Please, accept pull request fixing this issue

**Comment by tiangolo** [comment, marked as answer]
Hum, I see. The map between the path parameters in the path string and the schema should be based on the names, not on a specific order, OpenAPI doesn't have a concept of ordering on the schema of the path parameters. For autogenerated clients, I would expect them to generate methods that expect keyword arguments, not ordered ones. :thinking:
